drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+	 * Full Ip checksumm offload engine is configured in the hardware
+	 */
+	offload_needed = 1;
+
+	/* Enable the offload engine in the receive path */
+	synopGMAC_enable_rx_chksum_offload(gmacdev);
+
+	/* Default configuration, DMA drops the packets if error in encapsulated ethernet payload */
+	synopGMAC_rx_tcpip_chksum_drop_enable(gmacdev);
+#endif
+
+	for (i = 0; i < RECEIVE_DESC_SIZE; i++) {
+		skb = &rx_buf[NUMAKER_GMAC_INTF][i];
+		synopGMAC_set_rx_qptr(gmacdev, (u32)((u64)(skb->data) & NUMAKER_MASK_32),
+				      sizeof(skb->data), (u32)((u64)skb & NUMAKER_MASK_32));
+	}
+
+	for (i = 0; i < TRANSMIT_DESC_SIZE; i++) {
+		skb = &tx_buf[NUMAKER_GMAC_INTF][i];
+		synopGMAC_set_tx_qptr(gmacdev, (u32)((u64)(skb->data) & NUMAKER_MASK_32),
+				      sizeof(skb->data), (u32)((u64)skb & NUMAKER_MASK_32),
+				      offload_needed, 0);
+	}
+
+	synopGMAC_set_mac_address(NUMAKER_GMAC_INTF, mac_addr);
+
+	synopGMAC_clear_interrupt(gmacdev);
+
+	return status;
+}
+
+static int m_numaker_gmacdev_get_rx_buf(synopGMACdevice *gmacdev, uint16_t *len, uint8_t **buf)
+{
+	DmaDesc *rxdesc = gmacdev->RxBusyDesc;
+
+	LOG_DBG("---Start---");
+	if (synopGMAC_is_desc_owned_by_dma(rxdesc)) {
+		return -EIO;
+	}
+	if (synopGMAC_is_desc_empty(rxdesc)) {
+		return -ENODATA;
+	}
+
+	*len = synop_handle_received_data(NUMAKER_GMAC_INTF, buf);
+	if (*len <= 0) {
+		synopGMAC_enable_interrupt(gmacdev, DmaIntEnable);
+		return -ENOSPC; /* No available RX frame */
+	}
+
+	/* length of payload should be <= 1514 */
+	if (*len > (NU_ETH_MAX_FLEN - 4)) {
+		LOG_DBG("unexpected long packet length=%d, buf=0x%x", *len, (uint32_t)*buf);
+		*len = 0; /* Skip this unexpected long packet */
+	}
+	LOG_DBG("---End---");
+	return 0;
+}
+
+static void m_numaker_gmacdev_rx_next(synopGMACdevice *gmacdev)
+{
+	LOG_DBG("RX Next");
+	/* Already did in synop_handle_received_data
+	 * No-op at this stage
+	 * DmaDesc * rxdesc = (gmacdev->RxBusyDesc - 1);
+	 * rxdesc->status = DescOwnByDma;
+	 */
+}
+
+static void m_numaker_gmacdev_trigger_rx(synopGMACdevice *gmacdev)
+{
+	LOG_DBG("---Start---");
+
+	/* Enable the interrupt */
+	synopGMAC_enable_interrupt(gmacdev, DmaIntEnable);
+
+	/* Trigger RX DMA */
+	synopGMAC_enable_dma_rx(gmacdev);
+	synopGMAC_resume_dma_rx(gmacdev);
+	LOG_DBG("resume RX DMA");
+	LOG_DBG("---End---");
+}
+
+static void m_numaker_gmacdev_packet_rx(const struct device *dev)
+{
+	struct eth_numaker_data *data = dev->data;
+	synopGMACdevice *gmacdev = data->gmacdev;
+	uint8_t *buffer;
+	uint16_t len = 0;
+	struct net_pkt *pkt;
+	k_spinlock_key_t key;
+
+	/* Get exclusive access, use spin-lock instead of mutex in ISR */
+	key = k_spin_lock(&data->rx_frame_buf_lock);
+
+	/* Two approach: 1. recv all RX packets in one time.
+	 *               2. recv one RX and set pending interrupt for rx-next.
+	 */
+	while (1) {
+		/* get received frame */
+		if (m_numaker_gmacdev_get_rx_buf(gmacdev, &len, &buffer) != 0) {
+			break;
+		}
+
+		if (len == 0) {
+			break;
```

Comment:
is this an unexpected condition? may be worth a `LOG_WRN`

Reply:
OK, it's better to warn "No available RX frame".